### PR TITLE
bpo-31943: Add asyncio.Handle.cancelled() method

### PR DESCRIPTION
**Pull request by decaz**
It would be handy to access the loop's time the Handle's callback will be called at if you are using delayed calls with "delay" (not "when") argument.
Also it is useful to know whether the call was cancelled through the special attribute.

https://bugs.python.org/issue31943

### CODE REVIEW

**Comment by asvetlov** [comment]
@1st1 should we support it?
The change makes no harm but add version dependent behavior: user code should check if `when` attribute is present before accessing it.
On other side `call_later()` and family are low level API methods, most likely user should not call them from async functions.

**Comment by 1st1** [comment]
I'm OK with `.cancelled`, but I've no idea why we need `.when`. @decaz do you have a **real** use-case for it?

**Comment by 1st1** [comment]
BTW, `.cancelled` should be a method, not property, similar to `Future.cancelled()` method.

**Comment by decaz** [comment]
@1st1 yes, `.when` can be used to determine how much time is left before the delayed call is executed. For instance:
```python
seconds_left = handler.when - loop.time()
```
Now it's impossible as the time the handler will be called at is calculated internally.

**Comment by 1st1** [comment]
I specifically asked you about a real use case for ‘.when’. That means you should have told us about a real problem you encountered implementing a protocol or a server using asyncio, which was unsolvable without the ‘.when’ attribute.

We won’t add something just because there are hypothetical use cases for it.

**Comment by decaz** [comment]
@1st1 agree with you, just wanted to add such an attribute on the case if it will be convenient. I have rebased the branch.

File: Doc/library/asyncio-eventloop.rst
Comment (asvetlov) [inline Doc/library/asyncio-eventloop.rst:873]: Please add `.. versionadded:: 3.6`

**Review by asvetlov — CHANGES_REQUESTED** [review]

**Comment by asvetlov** [comment]
BTW `Handler` and `TimeHandler` are heavy used by asyncio core.
Maybe adding C implementation makes sense?
Even maybe supporting free lists for these classes might affect asyncio performance?

**Comment by 1st1** [comment]
> Maybe adding C implementation makes sense?

I can see it improving the performance somewhat.  But please hold off with the implementation before PEP 550 (or its successor) lands.

File: Doc/library/asyncio-eventloop.rst
Comment (1st1) [reply Doc/library/asyncio-eventloop.rst:873]: 3.7 :)

Comment (decaz) [reply Doc/library/asyncio-eventloop.rst:873]: OK, added. I hope it will be backported to 3.6?)

**Comment by decaz** [comment]
I have made the requested changes; please review again.

File: Doc/library/asyncio-eventloop.rst
Comment (1st1) [reply Doc/library/asyncio-eventloop.rst:873]: No. Asyncio is no longer provisional and this is not a bug fix.

**Review by asvetlov — APPROVED** [review]

**Comment by asvetlov** [comment]
Thanks for contribution!

**Pull request merged**
